### PR TITLE
Fix narrowing conversion in QIEDecoder

### LINKED ISSUE #1199: Narrowing conversion warnings by the compiler
**Describe the bug**
Inspired by the comment in https://github.com/LDMX-Software/ldmx-sw/pull/1198#issuecomment-1717288314 when I compile ldmx-sw I see the following narrowing conversions in ldmx-sw:

In `HcalGeometry.cxx`
```
/Users/tav/Documents/1Research/LDMX/ldmx-sw/DetDescr/src/DetDescr/HcalGeometry.cxx: In member function 'void ldmx::HcalGeometry::buildStripPositionMap()':
/Users/tav/Documents/1Research/LDMX/ldmx-sw/DetDescr/src/DetDescr/HcalGeometry.cxx:156:31: warning: narrowing conversion of 'section' from 'int' to 'unsigned int' [-Wnarrowing]
  156 |         const ldmx::HcalID id{section, layer, strip};
      |                               ^~~~~~~
/Users/tav/Documents/1Research/LDMX/ldmx-sw/DetDescr/src/DetDescr/HcalGeometry.cxx:156:40: warning: narrowing conversion of 'layer' from 'int' to 'unsigned int' [-Wnarrowing]
  156 |         const ldmx::HcalID id{section, layer, strip};
      |                                        ^~~~~
/Users/tav/Documents/1Research/LDMX/ldmx-sw/DetDescr/src/DetDescr/HcalGeometry.cxx:156:47: warning: narrowing conversion of 'strip' from 'int' to 'unsigned int' [-Wnarrowing]
  156 |         const ldmx::HcalID id{section, layer, strip};
      |                                               ^~~~~
/Users/tav/Documents/1Research/LDMX/ldmx-sw/DetDescr/src/DetDescr/HcalGeometry.cxx:211:57: warning: narrowing conversion of 'id.ldmx::HcalID::section()' from 'int' to 'unsigned int' [-Wnarrowing]
  211 |                       (getScintillatorLength({id.section(), 2, id.strip()}) -
      |                                               ~~~~~~~~~~^~
/Users/tav/Documents/1Research/LDMX/ldmx-sw/DetDescr/src/DetDescr/HcalGeometry.cxx:211:72: warning: narrowing conversion of 'id.ldmx::HcalID::strip()' from 'int' to 'unsigned int' [-Wnarrowing]
  211 |                       (getScintillatorLength({id.section(), 2, id.strip()}) -
      |                                                                ~~~~~~~~^~
/Users/tav/Documents/1Research/LDMX/ldmx-sw/DetDescr/src/DetDescr/HcalGeometry.cxx:238:57: warning: narrowing conversion of 'id.ldmx::HcalID::section()' from 'int' to 'unsigned int' [-Wnarrowing]
  238 |                       (getScintillatorLength({id.section(), 2, id.strip()}) -
      |                                               ~~~~~~~~~~^~
/Users/tav/Documents/1Research/LDMX/ldmx-sw/DetDescr/src/DetDescr/HcalGeometry.cxx:238:72: warning: narrowing conversion of 'id.ldmx::HcalID::strip()' from 'int' to 'unsigned int' [-Wnarrowing]
  238 |                       (getScintillatorLength({id.section(), 2, id.strip()}) -
```

and in `HgcrocEmulator.cxx`

```
/Users/tav/Documents/1Research/LDMX/ldmx-sw/Tools/src/Tools/HgcrocEmulator.cxx: In member function 'std::vector<ldmx::HgcrocDigiCollection::Sample> ldmx::HgcrocEmulator::noiseDigi(const int&, const double&) const':
/Users/tav/Documents/1Research/LDMX/ldmx-sw/Tools/src/Tools/HgcrocEmulator.cxx:235:17: warning: narrowing conversion of 'pedestal' from 'double' to 'int' [-Wnarrowing]
  235 |     int adc_tm1{pedestal};
      |                 ^~~~~~~~
/Users/tav/Documents/1Research/LDMX/ldmx-sw/Tools/src/Tools/HgcrocEmulator.cxx:240:24: warning: narrowing conversion of '(pedestal + (((const ldmx::HgcrocEmulator*)this)->ldmx::HgcrocEmulator::noise((* & channel)) / gain)' from 'double' to 'int' [-Wnarrowing]
  240 |     int adc_t{pedestal + noise(channel)/gain};
      |               ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
```

I believe this second one is responsible for the msg @bryngemark referred too.

**To Reproduce**

Just compiling with the latest gcc version

### PR DESCRIPTION
**Pull request by tvami**
I am updating _ldmx-sw_, here are the details.

### What are the issues that this addresses?
Resolves the last part of https://github.com/LDMX-Software/ldmx-sw/issues/1199
by statis casting.

## Check List
- [x] I successfully compiled _ldmx-sw_ with my developments
- [x] I ran my developments and the following shows that they are successful.

Resolves https://github.com/LDMX-Software/ldmx-sw/issues/1199

### CODE REVIEW

File: TrigScint/src/TrigScint/QIEDecoder.cxx [hunk 1]
```diff
@@ -63,15 +63,6 @@ void QIEDecoder::produce(framework::Event &event) {
   ldmx_log(debug) << "QIEDecoder: produce() starts! Event number: "
                   << event.getEventHeader().getEventNumber();
 
-  //	ldmx::EventHeader *eh = (ldmx::EventHeader*)event.getEventHeaderPtr();
-  // eh->setIsRealData(isRealData_); //doesn't help
-
-  /*   // comment for now, requires pushing change ot EventHeader
-  event.getEventHeader().setIsRealData(isRealData_);
```
Comment (bryngemark) [inline TrigScint/src/TrigScint/QIEDecoder.cxx:70]: I'd rather leave this in, still commented; we want to be able to set this somehow. Or we remove it but then make an issue for having a way to indicate that something is real data. (But you see the logic, if it weren't still there, I'd have forgotten about it...)

**Review by bryngemark — COMMENTED** [review]
Overall this looks good to me, but since there is no CI test for testbeam data stuff, have you tested this manually or just ticked the box?

File: TrigScint/src/TrigScint/QIEDecoder.cxx (hunk 1 above)
Comment (tvami) [reply TrigScint/src/TrigScint/QIEDecoder.cxx:70]: I'd prefer an issue, given that the problem is bigger then just the QIEDecoder

**Comment by tvami** [comment]
> have you tested this manually

I compiled and run the configs in the validation. What should I have run to test it on testbeam data?

**Comment by tvami** [comment]
> > have you tested this manually
> 
> I compiled and run the configs in the validation. What should I have run to test it on testbeam data?

Is this it? https://github.com/LDMX-Software/ldmx-sw/blob/trunk/TrigScint/exampleConfigs/runQIEDecode.py

File: TrigScint/src/TrigScint/QIEDecoder.cxx (hunk 1 above)
Comment (tvami) [reply TrigScint/src/TrigScint/QIEDecoder.cxx:70]: Here you go: https://github.com/LDMX-Software/ldmx-sw/issues/1428
The implementation doesnt seem to complicated, we just need to add another function to the event header right? If yes, I can try to do this next.

**Comment by tvami** [comment]
Well I ran that file, ran into an issue https://github.com/LDMX-Software/ldmx-sw/issues/1427 and fixed it https://github.com/LDMX-Software/ldmx-sw/pull/1429

Now when I run I get
```
---- LDMXSW: Loading configuration --------
[2024-09-02 10:58:03.478170] [0x00007f0bce616bc0] [info]    In configure, got parameters:
output_collection = decodedQIEUp
input_collection = QIEstreamUp
input_pass_name  = raw
channel_map_file = TrigScint/data/channelMap_LYSOback_plasticFront_12-to-16channels_rotated180.txt
number_channels  = 16
number_time_samples  = 1
is_real_data  = true
verbose          = true
[2024-09-02 10:58:03.479167] [0x00007f0bce616bc0] [info]    0  chID 0
[2024-09-02 10:58:03.479211] [0x00007f0bce616bc0] [info]    1  chID 2
[2024-09-02 10:58:03.479244] [0x00007f0bce616bc0] [info]    2  chID 10
[2024-09-02 10:58:03.479265] [0x00007f0bce616bc0] [info]    3  chID 4
[2024-09-02 10:58:03.479294] [0x00007f0bce616bc0] [info]    4  chID 3
[2024-09-02 10:58:03.479322] [0x00007f0bce616bc0] [info]    5  chID 1
[2024-09-02 10:58:03.479352] [0x00007f0bce616bc0] [info]    6  chID 5
[2024-09-02 10:58:03.479376] [0x00007f0bce616bc0] [info]    7  chID 11
[2024-09-02 10:58:03.479401] [0x00007f0bce616bc0] [info]    8  chID 15
[2024-09-02 10:58:03.479423] [0x00007f0bce616bc0] [info]    9  chID 6
[2024-09-02 10:58:03.479446] [0x00007f0bce616bc0] [info]    10  chID 14
[2024-09-02 10:58:03.479467] [0x00007f0bce616bc0] [info]    11  chID 8
[2024-09-02 10:58:03.479488] [0x00007f0bce616bc0] [info]    12  chID 7
[2024-09-02 10:58:03.479510] [0x00007f0bce616bc0] [info]    13  chID 13
[2024-09-02 10:58:03.479534] [0x00007f0bce616bc0] [info]    14  chID 9
[2024-09-02 10:58:03.479558] [0x00007f0bce616bc0] [info]    15  chID 12
[2024-09-02 10:58:03.479746] [0x00007f0bce616bc0] [info]    15  chID 12
---- LDMXSW: Configuration load complete  --------
---- LDMXSW: Starting event processing --------

 [ Process ] 1 : Opening file unpacked_ldmx_captan_out_09-04-2022_09-36-04__75_reformat_30timeSamplesFrom0_linearize_hits_clusters.root
Info in <TBranchElement::InitializeOffsets>: TTree created with an older schema, some data might not be copied in 'slow-cloning' mode; fast-cloning should have the correct result. 'timeSinceSpill_' is missing when constructing the branch 'EventHeader'. 
 [ Process ] 1 : Got new run header from 'out.root' ...
RunHeader { run: 1, numTries: 0, detectorName: ldmx-hcal-prototype-v1.0, description: 
  intParameters: 
    RandomNumberMasterSeed[raw] = 1
  floatParameters: 
  stringParameters: 
}
 [ Process ] 1 : Closing file unpacked_ldmx_captan_out_09-04-2022_09-36-04__75_reformat_30timeSamplesFrom0_linearize_hits_clusters.root
---- LDMXSW: Event processing complete  --------
```
which makes me think that it's technically working. If you have a better input file I can run on that too. Maybe I should tag @hherde too.

**Review by tomeichlersmith — APPROVED** [review]

**Comment by tvami** [comment]
OK so running 
```
ldmx recompFire TrigScint/exampleConfigs/runQIEDecode.py /fs/ddn/sdf/group/ldmx/data/TS-data/test_stand_data/rootFiles/unpacked_ldmx_captan_out_09-04-2022_09-36-04__75_reformat_30timeSamplesFrom0_linearize_hits_clusters.root out.root raw 1 TrigScint/data/channelMap_LYSOback_plasticFront_12-to-16channels_rotated180.txt
```
works, which makes me think that this is enough to merge this.

**Comment by bryngemark** [comment]
I can't tell from the output if what you ran actually did anything. But I'm rather confident your change is fine. Making an issue for setting up tests for our data workflow is a good outcome of this PR.

**Review by bryngemark — APPROVED** [review]

**Pull request merged**